jax/_src/custom_derivatives.py

```diff
@@ -1424,53 +1424,56 @@ def linear_call(fun: Callable,
                                 (residual_args, linear_args), {})),
       t_in_tree)
 
-  t_jaxpr, t_consts = _initial_style_jaxpr(t, (*res_avals, *out_avals))
-  t_jaxpr_closed = _close_jaxpr(t_jaxpr)
-
-  if t_out_tree() != lin_tree:
-    raise TypeError(
-        'transpose output pytree structure must match that of linear inputs, '
-        f'got output structure {t_out_tree()} '
-        f'and input structure {lin_tree}.')
+  @pe._memoize
+  def transpose_thunk():
+    t_jaxpr, t_consts = _initial_style_jaxpr(t, (*res_avals, *out_avals))
+    return _close_jaxpr(t_jaxpr), t_consts
 
-  out = linear_call_p.bind(*f_consts, *t_consts, *operands_res, *operands_lin,
+  out = linear_call_p.bind(*f_consts, *operands_res, *operands_lin,
                            callee=f_jaxpr_closed,
-                           transpose=t_jaxpr_closed,
+                           transpose_thunk=transpose_thunk,
                            num_callee_consts=len(f_consts),
-                           num_transpose_consts=len(t_consts),
                            num_res=len(operands_res))
 
+  try:
+    if t_out_tree() != lin_tree:
+      raise TypeError(
+          'transpose output pytree structure must match that of linear inputs, '
+          f'got output structure {t_out_tree()} '
+          f'and input structure {lin_tree}.')
+  except lu.StoreException:
+    pass
```

Comment:
Does this mean that we might miss out on checking `t_out_tree`, in a situation where previously we would have checked it?

Reply:
Good question! This store won't be populated until `transpose_thunk` is called. This means that we won't get any checking unless we transpose the function. I'm not so concerned about those cases, but it might be a good idea to audit this check to make sure that it gets executed when we want it to. I could also move the check into the body of `transpose_thunk`. I'll add a test tomorrow!

Reply:
You were right! This wasn't in a good spot. I've moved this check into `transpose_thunk`, so that it now raises properly when the transpose is called. This is a change of behavior because previously the error would have been raised on the fwd pass, but I think this is fine. WDYT?

Reply:
Yeah, the point at which the check happens must move, because we're deferring evaluation. But at least it isn't skipped now. Looks good to me.